apollo-federation/src/sources/connect/models/validation.rs

```diff
@@ -569,100 +575,92 @@ fn parse_url(value: &Node<Value>, coordinate: &str, sources: &SourceMap) -> Resu
+            let pair_coordinate = &directive_http_header_coordinate(
+                directive_name,
+                argument_name,
+                object,
+                field,
+            );
 
-                    if key == &name!("as") {
-                        if has_as_or_value {
-                            // TODO: update this for correct error handling
-                            errors.push(Message {
-                                code: Code::InvalidHttpHeaderPair,
-                                message: format!("{pair_coordinate} uses both `as` and `value` keys together. Please choose only one."),
-                                locations: Location::from_node(value.location(), source_map)
-                                    .into_iter()
-                                    .collect(),
-                            });
-                        } else if key == &name!("as") {
-                            has_as_or_value = true;
-                        }
-                    }
-                } else if key == &name!("value") {
-                    if let Some(err) =
-                        validate_header_value(value, pair_coordinate, source_map).err()
-                    {
-                        errors.extend(err);
-                    }
-                    if has_as_or_value {
-                        // TODO: update this for correct error handling
+            let name_arg = arg_pairs.iter().find_map(|(key, value)| (key == &name!("name")).then_some(value));
+            let as_arg = arg_pairs.iter().find_map(|(key, value)| (key == &name!("as")).then_some(value));
+            let value_arg = arg_pairs.iter().find_map(|(key, value)| (key == &name!("value")).then_some(value));
+
+            // validate `name`
+            if let Some(name_value) = name_arg {
+                if let Some(err) = validate_header_name(&name!("name"), name_value, pair_coordinate, source_map).err() {
+                    errors.push(err);
+                }
+            } else {
+                // `name` must be provided
+                errors.push(Message {
+                    code: Code::HttpHeaderNameClash,
+                    message: format!("{pair_coordinate} must include a `name` value."),
+                    // TODO: get this closer to the pair
+                    locations: Location::from_node(headers.location(), source_map)
+                        .into_iter()
+                        .collect(),
+                });
+            }
+
+            // validate `as`
+            if let Some(as_value) = as_arg {
+                if let Some(err) = validate_header_name(&name!("as"), as_value, pair_coordinate, source_map).err() {
+                    errors.push(err);
+                }
+            }
+
+            if let (Some(as_arg), Some(name_arg)) = (as_arg, name_arg) {
+                if let (Some(as_value), Some(name_value)) = (as_arg.as_str(), name_arg.as_str()) {
+                    if as_value == name_value {
```

Comment:
i didn't even think of this one! it makes sense as it's probably an accident, but you could argue that `{ name: "x-foo" }` is really just a convenience for `{ name: "x-foo", as: "x-foo" }`.

when i first read this, i thought it was validating that `name`s are unique across header objects.

maybe this is actually a bit more complicated than i thought.

```graphql
@source(
  name: "json",
  http: { 
    baseURL: "http://whatever"
    headers: [
      { name: "x-foo", as: "x-bar" }
      { name: "x-foo", values: ["baz"] }
    ]
)
```
At runtime, this should result in a valid map:
```json
{ "x-bar": req.headers["x-foo"], "x-foo": ["baz"] }
```

now i'm wondering if it should actually be
```
    headers: [
      { name: "x-bar", from: "x-foo" }
      { name: "x-foo", values: ["baz"] }
    ]
```

where `{ name: "x-bar" }` is a convenience for `{ name: "x-bar", from: "x-bar" }`.

this would make the validation rule simpler: all `name` attributes must be unique.